File: src/aoNewsletter2/Controllers/blockClass.vb

Option Explicit On
Option Strict On

Imports System
Imports System.Collections.Generic
Imports System.Text
Imports Contensive.BaseClasses

Namespace newsletter2
    ''' <summary>
    ''' A replacement for cp.block using htmlAgility
    ''' </summary>
    Public Class blockClass
        '
        Dim layout2 As HtmlAgilityPack.HtmlDocument
        '
        Public Sub load(html As String)
            layout2 = New HtmlAgilityPack.HtmlDocument
            layout2.LoadHtml(html)
        End Sub
        '
        Public Sub prepend(html As String)
            layout2.Load(html & getHtml())
        End Sub
        '
        Public Sub append(html As String)
            layout2.Load(getHtml() & html)
        End Sub
        '
        Public Function getClassInner(findClass As String) As String
            Dim node As HtmlAgilityPack.HtmlNode = layout2.DocumentNode.SelectSingleNode("//*[@class='" & findClass & "']")
            If Not (node Is Nothing) Then
                Return node.InnerHtml()
            End If
            Return String.Empty
        End Function
        '
        Public Function getClassOuter(findClass As String) As String
            Dim node As HtmlAgilityPack.HtmlNode = layout2.DocumentNode.SelectSingleNode("//*[@class='" & findClass & "']")
            If Not (node Is Nothing) Then
                Return node.OuterHtml()
            End If
            Return String.Empty
        End Function
        '
        Public Sub setClassInner(findClass As String, replacement As String)
            Dim nodes As HtmlAgilityPack.HtmlNodeCollection = layout2.DocumentNode.SelectNodes("//*[@class='" & findClass & "']")
            If Not (nodes Is Nothing) Then
                For Each node As HtmlAgilityPack.HtmlNode In nodes
                    node.InnerHtml = replacement
                Next
            End If
        End Sub
        '
        Public Sub setClassOuter(findClass As String, replacement As String)
            Dim nodes As HtmlAgilityPack.HtmlNodeCollection = layout2.DocumentNode.SelectNodes("//*[@class='" & findClass & "']")
            If Not (nodes Is Nothing) Then
                For Each node As HtmlAgilityPack.HtmlNode In nodes
                    Dim newNode As HtmlAgilityPack.HtmlNode = HtmlAgilityPack.HtmlNode.CreateNode(replacement)
                    node.ParentNode.ReplaceChild(newNode, node)
                Next
            End If
        End Sub
        '
        Public Function getHtml() As String
            If Not layout2 Is Nothing Then
                If Not (layout2.DocumentNode Is Nothing) Then
                    Return layout2.DocumentNode.OuterHtml
                End If
            End If
            Return String.Empty
        End Function
    End Class
End Namespace

File: server/aoNewsletter2/Controllers/blockClass.cs


namespace Contensive.Addons.Newsletter.Controllers {
    /// <summary>
    /// A replacement for cp.block using htmlAgility
    /// </summary>
    public class BlockClass {
        // 
        private HtmlAgilityPack.HtmlDocument layout2;
        // 
        public void load(string html) {
            layout2 = new HtmlAgilityPack.HtmlDocument();
            layout2.LoadHtml(html);
        }
        // 
        public void prepend(string html) {
            layout2.Load(html + getHtml());
        }
        // 
        public void append(string html) {
            layout2.Load(getHtml() + html);
        }
        // 
        public string getClassInner(string findClass) {
            var node = layout2.DocumentNode.SelectSingleNode($"//*[contains(concat(' ', normalize-space(@class), ' '), ' {findClass} ')]");
            if (node is not null) {
                return node.InnerHtml;
            }
            return string.Empty;
        }
        // 
        public string getClassOuter(string findClass) {
            var node = layout2.DocumentNode.SelectSingleNode($"//*[contains(concat(' ', normalize-space(@class), ' '), ' {findClass} ')]");
            if (node is not null) {
                return node.OuterHtml;
            }
            return string.Empty;
        }
        // 
        public void setClassInner(string findClass, string replacement) {
            var nodes = layout2.DocumentNode.SelectNodes($"//*[contains(concat(' ', normalize-space(@class), ' '), ' {findClass} ')]");
            if (nodes is not null) {
                foreach (HtmlAgilityPack.HtmlNode node in nodes)
                    node.InnerHtml = replacement;
            }
        }
        // 
        public void setClassOuter(string findClass, string replacement) {
            var nodes = layout2.DocumentNode.SelectNodes($"//*[contains(concat(' ', normalize-space(@class), ' '), ' {findClass} ')]");
            if (nodes is not null) {
                foreach (HtmlAgilityPack.HtmlNode node in nodes) {
                    var newNode = HtmlAgilityPack.HtmlNode.CreateNode(replacement);
                    node.ParentNode.ReplaceChild(newNode, node);
                }
            }
        }
        // 
        public string getHtml() {
            if (layout2 is not null) {
                if (layout2.DocumentNode is not null) {
                    return layout2.DocumentNode.OuterHtml;
                }
            }
            return string.Empty;
        }
    }
}